File: 0067-add-binary/0067-add-binary.java
class Solution {
    public String addBinary(String a, String b) {
        if (b.length() > a.length()) {
            return addBinary(b, a);
        }
        StringBuilder result = new StringBuilder();
        int carry = 0;
        for (int i = 0; i < a.length(); ++i) {
            int sum = carry;
            if (a.charAt(a.length() - 1 - i) == '1') {
                sum++;
            }
            if (i < b.length() && b.charAt(b.length() - 1 - i) == '1') {
                sum++;
            }
            result.append(sum % 2);
            carry = sum / 2;
        }
        if (carry > 0) {
            result.append(carry);
        }
        result.reverse();
        return result.toString();
    }
}

File: 0067-add-binary/0067-add-binary.kt
class Solution {
    fun addBinary(a: String, b: String): String {
        val result = StringBuilder()
        var carry = 0
        for (i in 0 until maxOf(a.length, b.length)) {
            var d1 = 0
            if (i < a.length) {
                d1 = a[a.length - i - 1] - '0'
            }
            var d2 = 0
            if (i < b.length) {
                d2 = b[b.length - i - 1] - '0'
            }
            val d = d1 + d2 + carry
            result.append(d % 2)
            carry = d / 2
        }
        if (carry > 0) {
            result.append(carry)
        }
        result.reverse()
        return result.toString()
    }
}